### PR TITLE
integration of timestamp utils

### PR DESCRIPTION
**Pull request by Gautzilla**
This PR aims at integrating the new timestamp utils to the OSEkit codebase.

It might start to enlighten the code, notably in the `Dataset` class.

Some info which was previously printed might temporarily disappear, since the whole log workflow will be reworked in the near future.

### CODE REVIEW

**Review by mathieudpnt — APPROVED** [review]

**Comment by Gautzilla** [comment]
# Recent changes

Fixed some bugs, but are not definitive, but also temporarily raises an error if the timestamp.csv already contains OSmOSE-formatted timestamps but the user specifies the old datetime format.

## Bugs fixed

Concern the timestamps parsing: now, all supported strftime blocks (%.) should be usable.

## Timezone conflicts management

### Warning

A warning is raised if all of the following items are true:
- User's timestamps.csv contains tz-aware timestamps
- User's datetime_format includes the timezone (through either %z or %Z)
- `Dataset.build()` is called with a specified timezone which UTC offset differs from the one embedded in the user's timestamps

### Timezone conversion

In the case where a warning is raised, the timestamps are **converted** from the timezone contained within the user's timestamps.csv **to the specified timezone**.

example:

User timestamp.csv:
```csv
filename, timestamp
"audio_1.wav", "01-01-2000 12:00:00 +0200"
```

Build call:
```py
dataset.build(datetime_format = "%d-%m-%Y %H:%M:%S %z", timezone = "UTC")
```
User is warned because the `UTC` timezone passed to the build method doesn't match the +0200 timezone parsed from the timestamp.csv file.
The timestamps will be **converted** to UTC:


result timestamp.csv:
```csv
filename, timestamp
"audio_1.wav", "2020-01-01T10:00:00.0000+0000"
```

## Cases where no conflict is declared

### No timezone identifier in `datetime_format`

If the user doesn't specify the timezone identifier in the `datetime_format` argument, it will not be parsed (and the timestamp will be read as it is, in the timezone specified in the `timezone` argument):

User timestamp.csv:
```csv
filename, timestamp
"audio_1.wav", "01-01-2000 12:00:00 +0200"
```

Build call:
```py
dataset.build(datetime_format = "%d-%m-%Y %H:%M:%S ", timezone = "UTC") # NO %z IN datetime_format
```

result timestamp.csv:
```csv
filename, timestamp
"audio_1.wav", "2020-01-01T12:00:00.0000+0000"
```

### Timezones are different but have the same UTC offset

If the timezone passed to the build method and parsed from the timestamps.csv have similar UTC offset, no warning is raised:

User timestamp.csv:
```csv
filename, timestamp
"audio_1.wav", "01-01-2000 12:00:00 Europe/Madrid"
```

Build call:
```py
dataset.build(datetime_format = "%d-%m-%Y %H:%M:%S %Z ", timezone = "Europe/Malta") # timezone parsed with %Z, but it works the same for %z timezones, both in the timestamp.csv and in the timezone argument
```

result timestamp.csv:
```csv
filename, timestamp
"audio_1.wav", "2020-01-01T12:00:00.0000+0100"
```

# TODO

I still have to completely rewrite the `Dataset.build()` method to fully integrate these changes.

## Induced bugs

To build a dataset which timestamps.csv is **already formatted to the OSmOSE format**, the user **must not** provide another datetime_format. This didn't raise an error before, but it does now. 

The changes have not been implemented to the `Auxiliary` folder.

Both these two last points will be fixed right when I'll rewrite the `build()` method.

**Comment by Gautzilla** [comment]
This PR led to more substantial changes to the Dataset class, which I'm working on on the [refactor/dataset-build](https://github.com/Gautzilla/OSEkit/tree/refactor/dataset-build) branch on my fork. 

I cleaned up a lot while trying to keep everything working in the same way. I still have to:

- Resolve the way the original folder is found for non-built datasets in a clean way and move it into a single method
- clean up the way the file_metadata DataFrame is returned by the _build_audio(audio_path=audio_path, date_template=date_template) method: I might have to explode the build_audio method
- clean up the creation of the metadata.csv file

I might even move all the metadata-related things to a separate class, which might help keeping the logic to its own place.

**Comment by Gautzilla** [comment]
I realize this PR has already grown too big and covers a way larger scope than its original one...

Basically, it started with some small integration of the new timestamps utils, with added testing. Then I realized I was stuck by the old `build()` method, which I tried to simplify while keeping most of its original behaviour.

The whole data reshape thing discussed in #228 will imply further changes on all this, but this will still be a cleaner base on which to work on!

**Comment by Gautzilla** [comment]
# Changes made to the `Dataset.build()` method

## `resume_anomalies` is no more 🪦

### Building a dataset with providing an incomplete `timestmap.csv` will lead to a logged error:

```md
# logs.txt
	2024-11-29 12:25:12,700 - dataset.small_dataset - ERROR
7189.230405154906.wav has not been found in timestamp.csv
```

### Building a dataset with providing a `timestmap.csv` and missing files will lead to a logged error:

```md
# logs.txt
	2024-11-29 12:29:35,036 - dataset.small_dataset
7189.230405164906.wav is listed in timestamp.csv but hasn't be found.
```

## Automatic finding of the original folder has been removed

This was a bit error-prone, so now the original folder (the one containing the data used for building) is determined in this order:
- If the user specifies it as a parameter to the `build()` method (`dataset.build(original_folder = path)`, with `path` being either relative to the dataset path or absolute), it is chosen.
- If not, it is the **root** folder of the dataset.

## All files that are not used in the build process are **moved** to an `other` folder:

```md
# Before build

dataset
├── 7189.230405144906.wav
├── 7189.230405154906.wav
├── 7189.230405164906.wav
├── 7189.230405174906.wav
├── some_file.txt
└── foo
    ├── some_audio.wav
    └── bar
        ├── some_other_file.txt
        └── empty_folder
```

# After build

# log:
	2024-11-29 12:36:50,692 - dataset.small_dataset
Moved 3 file(s) to the 'other' folder.

# folder structure:

dataset
├── data
│   └── audio
│       └── 3600_128000
│           ├── 7189.230405144906.wav
│           ├── 7189.230405154906.wav
│           ├── 7189.230405164906.wav
│           ├── 7189.230405174906.wav    
│           ├── file_metadata.csv
│           ├── metadata.csv
│           └── timestamp.csv
├── log
│   └── logs.log
└── other
    ├── some_file.txt
    └── foo
        ├── some_audio.wav
        └── bar
            └── some_other_file.txt
```
Note: empty folders are deleted rather than being moved to the other folder.

File: src/OSmOSE/Dataset.py [hunk 1]
```diff
@@ -1,35 +1,30 @@
+from __future__ import annotations
```
Comment (mathieudpnt) [inline src/OSmOSE/Dataset.py:1]: what does this import stand for ?

Comment (Gautzilla) [reply src/OSmOSE/Dataset.py:1]: It's just a [matter of compatibility](https://docs.astral.sh/ruff/rules/future-required-type-annotation/) with Python 3.10- versions for [PEP 604 type unions](https://peps.python.org/pep-0604/).

**Review by mathieudpnt — APPROVED** [review]
Tried it with and without user-defined begin/end datetimes, worked smoothly for me

**Pull request merged**